python/lsst/meas/algorithms/reserveSourcesTask.py

```diff
@@ -0,0 +1,183 @@
+# see <https://www.lsstcorp.org/LegalNotices/>.
+#
+from __future__ import absolute_import, division, print_function
+
+from builtins import zip
+
+import numpy as np
+
+from lsst.pex.config import Config, Field
+from lsst.pipe.base import Task, Struct
+
+import lsst.afw.table
+
+__all__ = ["ReserveSourcesConfig", "ReserveSourcesTask",]
+
+class ReserveSourcesConfig(Config):
+    """Configuration for reserving sources"""
+    fraction = Field(dtype=float, default=0.0,
+                     doc="Fraction of candidates to reserve from fitting; none if <= 0")
+    seed = Field(dtype=int, default=1,
+                 doc=("This number will be added to the exposure ID to set the random seed for "
+                      "reserving candidates"))
+
+
+class ReserveSourcesTask(Task):
+    """Reserve sources from analysis
+
+    We randomly select a fraction of sources that will be reserved
+    from analysis. This allows evaluation of the quality of model fits
+    using sources that were not involved in the fitting process.
+
+    Constructor parameters
+    ----------------------
+    columnName : `str`, required
+        Name of flag column to add; we will suffix this with "_reserved".
+    schema : `lsst.afw.table.Schema`, required
+        Catalog schema.
+    doc : `str`
+        Documentation for column to add.
+    config : `ReserveSourcesConfig`
+        Configuration.
+    """
+    ConfigClass = ReserveSourcesConfig
+    _DefaultName = "reserveSources"
+
+    def __init__(self, columnName=None, schema=None, doc=None, **kwargs):
+        Task.__init__(self, **kwargs)
+        assert columnName is not None, "columnName not provided"
+        assert schema is not None, "schema not provided"
+        self.columnName = columnName
+        self.key = schema.addField(self.columnName + "_reserved", type="Flag", doc=doc)
+
+    def run(self, sources, prior=None, expId=0):
+        """Select sources to be reserved
+
+        Reserved sources will be flagged in the catalog, and we will return
+        boolean arrays that identify the sources to be reserved from and
+        used in the analysis.
+
+        Parameters
+        ----------
+        sources : `lsst.afw.table.Catalog` or `list` of `lsst.afw.table.Record`
+            Sources from which to select some to be reserved.
+        prior : `numpy.ndarray` of type `bool`, optional
+            Prior selection of sources. Should have the same length as
+            `sources`. If set, we will only consider for reservation sources
+            that are flagged `True` in this array.
+        expId : `int`
+            Exposure identifier; used for seeding the random number generator.
+
+        Return struct contents
+        ----------------------
+        reserved : `numpy.ndarray` of type `bool`
+            Sources to be reserved are flagged `True` in this array.
+        use : `numpy.ndarray` of type `bool`
+            Sources the user should use in analysis.
```

Comment:
The distinction between `reserved` and `use` is not that clear here. What "analysis"?